src/concurrency/local_epoch.cpp

```diff
@@ -0,0 +1,118 @@
+//===----------------------------------------------------------------------===//
+//
+//                         Peloton
+//
+// centralized_epoch_manager.cpp
+//
+// Identification: src/concurrency/local_epoch.cpp
+//
+// Copyright (c) 2015-16, Carnegie Mellon University Database Group
+//
+//===----------------------------------------------------------------------===//
+
+#include "concurrency/decentralized_epoch_manager.h"
+
+
+namespace peloton {
+namespace concurrency {
+
+  bool LocalEpoch::EnterEpoch(const uint64_t epoch_id) {
+    epoch_lock_.Lock();
+    // if this thread is never used or GC'd
+    if (epoch_lower_bound_ == UINT64_MAX) {
+      epoch_lower_bound_ = epoch_id - 1;
+    } else if (epoch_lower_bound_ >= epoch_id) {
+    // epoch_lower_bound_ has already been updated by the GC.
+    // have to grab a newer epoch_id.
+      epoch_lock_.Unlock();
+      return false;
+    } 
+    // check whether the corresponding epoch exists.
+    if (epoch_map_.find(epoch_id) == epoch_map_.end()) {
```

Comment:
Store the result of find(epoch_id) in a local variable and use it if it's valid.